cogito/put_test.go

```diff
@@ -0,0 +1,220 @@
+package cogito_test
+
+import (
+	"bytes"
+	"errors"
+	"io"
+	"strings"
+	"testing"
+	"testing/iotest"
+
+	"github.com/Pix4D/cogito/cogito"
+	"github.com/Pix4D/cogito/testhelp"
+	"github.com/hashicorp/go-hclog"
+	"gotest.tools/v3/assert"
+)
+
+func TestPutSuccess(t *testing.T) {
+	type testCase struct {
+		name    string
+		in      cogito.PutInput
+		wantOut cogito.Output
+	}
+
+	test := func(t *testing.T, tc testCase) {
+		in := bytes.NewReader(toJSON(t, tc.in))
+		var out bytes.Buffer
+		log := hclog.NewNullLogger()
+		inputDir := testhelp.MakeGitRepoFromTestdata(t, "testdata/one-repo/a-repo",
+			testhelp.HttpsRemote("the-owner", "the-repo"), "dummySHA", "dummyHead")
+
+		err := cogito.Put(log, in, &out, []string{inputDir})
+
+		assert.NilError(t, err)
+		var have cogito.Output
+		fromJSON(t, out.Bytes(), &have)
+		assert.DeepEqual(t, have, tc.wantOut)
+	}
+
+	baseSource := cogito.Source{
+		Owner:       "the-owner",
+		Repo:        "the-repo",
+		AccessToken: "the-token",
+	}
+
+	baseParam := cogito.PutParams{State: cogito.StatePending}
+
+	testCases := []testCase{
+		{
+			name: "returns correct version and metadata",
+			in: cogito.PutInput{
+				Source: baseSource,
+				Params: baseParam,
+			},
+			wantOut: cogito.Output{
+				Version:  cogito.DummyVersion,
+				Metadata: []cogito.Metadata{{Name: cogito.KeyState, Value: cogito.StatePending}},
+			},
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) { test(t, tc) })
+	}
+}
+
+func TestPutPipelineValidationFailure(t *testing.T) {
+	type testCase struct {
+		name     string
+		putInput cogito.PutInput
+		wantErr  string
+	}
+
+	test := func(t *testing.T, tc testCase) {
+		log := hclog.NewNullLogger()
+		in := bytes.NewReader(toJSON(t, tc.putInput))
+
+		err := cogito.Put(log, in, io.Discard, []string{"dummy-dir"})
+
+		assert.Error(t, err, tc.wantErr)
+	}
+
+	baseSource := cogito.Source{
+		Owner:       "the-owner",
+		Repo:        "the-repo",
+		AccessToken: "the-token",
+	}
+
+	testCases := []testCase{
+		{
+			name: "missing keys",
+			putInput: cogito.PutInput{
+				Source: cogito.Source{},
+			},
+			wantErr: "put: source: missing keys: owner, repo, access_token",
+		},
+		{
+			name: "invalid log_level",
+			putInput: cogito.PutInput{
+				Source: mergeStructs(baseSource, cogito.Source{LogLevel: "pippo"}),
+			},
+			wantErr: "put: source: invalid log_level: pippo",
+		},
+		{
+			name: "invalid params",
+			putInput: cogito.PutInput{
+				Source: baseSource,
+				Params: cogito.PutParams{State: "burnt-pizza"},
+			},
+			wantErr: "put: params: invalid build state: burnt-pizza",
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) { test(t, tc) })
+	}
+
+}
+
+func TestPutProtocolFailure(t *testing.T) {
```

Comment:
These tests validate "should not happen" situation: resource protocol errors introduced by Concourse.